runtime/expr/function/types.go

```diff
@@ -107,6 +108,78 @@ func (i *Is) Call(_ zed.Allocator, args []zed.Value) *zed.Value {
 	return zed.False
 }
 
+type HasError struct {
+	cached map[int]bool
+}
+
+func NewHasError() *HasError {
+	return &HasError{
+		cached: make(map[int]bool),
+	}
+}
+
+func (h *HasError) Call(_ zed.Allocator, args []zed.Value) *zed.Value {
+	v := args[0]
+	if yes, _ := h.hasError(v.Type, v.Bytes); yes {
+		return zed.True
+	}
+	return zed.False
+}
+
+func (h *HasError) hasError(t zed.Type, b zcode.Bytes) (bool, bool) {
```

Comment:
I think you could simplify this if you used `Value.Walk`.

Reply:
I had considered this but there was some case that I thought had made it unworkable. Let me look again.

Reply:
Using walk is tougher because I need to keep track of whether a part of value tree is cacheable or not.